examples/ota-provider-app/ota-provider-common/OTAProviderExample.cpp

```diff
@@ -91,26 +93,88 @@ void OTAProviderExample::SetOTAFilePath(const char * path)
     }
 }
 
+void OTAProviderExample::SetOTACandidates(std::vector<OTAProviderExample::DeviceSoftwareVersionModel> candidates)
+{
+    mCandidates = std::move(candidates);
+}
+
+void OTAProviderExample::SetOTACandidates(OTAProviderExample::DeviceSoftwareVersionModel entry)
+{
+    mCandidates.push_back(entry);
+}
+
+static bool compareSoftwareVersions(const OTAProviderExample::DeviceSoftwareVersionModel & a,
+                                    const OTAProviderExample::DeviceSoftwareVersionModel & b)
+{
+    return (a.softwareVersion < b.softwareVersion);
+}
+
+bool OTAProviderExample::SelectOTACandidate(const uint16_t requestorVendorID, const uint16_t requestorProductID,
+                                            const uint32_t requestorSoftwareVersion,
+                                            OTAProviderExample::DeviceSoftwareVersionModel & finalCandidate)
+{
+    bool candidateFound = false;
+    std::sort(mCandidates.begin(), mCandidates.end(), compareSoftwareVersions);
+    for (auto candidate : mCandidates)
+    {
+        // VendorID and ProductID will be the primary key when querying
+        // the DCL servers. If not we can add the vendor/product ID checks here.
+        if (candidate.softwareVersionValid &&
+            requestorSoftwareVersion < candidate.softwareVersion &&
+            requestorSoftwareVersion >= candidate.minApplicableSoftwareVersion &&
+            requestorSoftwareVersion <= candidate.maxApplicableSoftwareVersion)
+        {
+            candidateFound = true;
```

Comment:
nit: I wonder if you can make `finalCandidate` Optional/Nullable such that you won't need to set a bool to return. The caller can check if candidate.HasValue()

Reply:
Nevermind, I see that below, you use the return value in multiple places.